src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+        Selection ret{{}};
+        for (const auto& v : values_) {
+            ret = ret | np.matchAttributeValues(attribute_, v);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString(attribute_, values_);
+    }
+
+  private:
+    std::string attribute_;
+    std::vector<T> values_;
+};
+
+class NodeSetBasicPopulation: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicPopulation(std::string name)
+        : name_(std::move(name)) {}
+
+    Selection materialize(const NodeSets& /* unused */, const NodePopulation& np) const final {
+        if (name_ == np.name()) {
+            return np.selectAll();
+        }
+
+        return Selection{{}};
+    }
+
+    std::string toJSON() const final {
+        const std::vector<std::string> values = {name_};
+        return toString("population", values);
+    }
+
+  private:
+    std::string name_;
+};
+
+class NodeSetBasicNodeIds: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicNodeIds(Selection::Values&& values)
+        : values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */,
+                          const NodePopulation& /* unused */) const final {
+        return Selection::fromValues(values_.begin(), values_.end());
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", values_);
+    }
+
+  private:
+    Selection::Values values_;
+};
+
+using CompoundTargets = std::vector<std::string>;
+class NodeSetCompoundRule: public NodeSetRule
+{
+  public:
+    NodeSetCompoundRule(std::string attribute, const CompoundTargets& targets)
+        : attribute_(std::move(attribute))
+        , targets_(targets) {}
+
+    Selection materialize(const NodeSets& ns, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& target : targets_) {
+            ret = ret | ns.materialize(target, np);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", targets_);
+    }
+
+  private:
+    std::string attribute_;
+    CompoundTargets targets_;
+};
+
+int64_t get_integer_or_throw(const json& el) {
+    auto v = el.get<double>();
+    if (std::floor(v) != v) {
+        throw SonataError("Only allowed integers in node set rules");
+    }
+    return static_cast<int64_t>(v);
+}
+
+NodeSetRules _dispatch_node(const json& contents) {
+    NodeSetRules ret;
+
+    for (auto& el : contents.items()) {
+        const std::string& attribute = el.key();
+        if (el.value().is_number()) {
+            if (attribute == "population") {
+                throw SonataError("'population' must be a string");
+            }
+
+            int64_t v = get_integer_or_throw(el.value());
+            if (attribute == "node_id") {
+                if (v < 0) {
+                    throw SonataError("'node_id' must be positive");
+                }
```

Comment:
this will depends if you are using the "or" for the "node_id" key like I do in snap

Reply:
Not sure what you mean; shouldn't `node_id` be positve?

Reply:
If the 'node_id' : [] act like a "or" then you could have (imao) [-5, 5, 6, -3] --> [5, 6] without raising. 
So yes node_id should be positive so you will never select the negative ids but I am not sure you should raise

Reply:
I think raising is appropriate: if someone puts a negative number, they should be warned, b/c it's probably a mistake.

Reply:
ok fair enough. This is just a behavior we should share among tools